internal/template/utilization_functions.go

```diff
@@ -72,69 +73,108 @@ scalar(
   * 100, 0.0001)
 )`
 
-	return renderUtilization(memoryUtilizationQueryTemplate, data, labelArgs...)
+	return renderUtilization(data, labelMatchers, memoryUtilizationQueryTemplate)
 }
 
-func cpuRequests(data MetricData, labelArgs ...string) (string, error) {
+func cpuRequests(data MetricData, labelMatchers ...string) (string, error) {
 	cpuResourcesQueryTemplate := `
 scalar(
   sum(
     avg_over_time(kube_pod_container_resource_requests_cpu_cores[{{ .Range }}])
     *
-    on (pod) group_left kube_pod_labels{{ .Labels }}
+    on (pod) group_left kube_pod_labels{{ .MetricSelector }}
   )
 )`
 
-	return renderUtilization(cpuResourcesQueryTemplate, data, labelArgs...)
+	return renderUtilization(data, labelMatchers, cpuResourcesQueryTemplate)
 }
 
-func memoryRequests(data MetricData, labelArgs ...string) (string, error) {
+func memoryRequests(data MetricData, labelMatchers ...string) (string, error) {
 	memoryResourcesQueryTemplate := `
 scalar(
   sum(
     avg_over_time(kube_pod_container_resource_requests_memory_bytes[{{ .Range }}])
     *
-    on (pod) group_left kube_pod_labels{{ .Labels }}
+    on (pod) group_left kube_pod_labels{{ .MetricSelector }}
   )
 )`
 
-	return renderUtilization(memoryResourcesQueryTemplate, data, labelArgs...)
+	return renderUtilization(data, labelMatchers, memoryResourcesQueryTemplate)
 }
 
-func renderUtilization(query string, data MetricData, labelArgs ...string) (string, error) {
-	tmpl := template.Must(template.New("query").Parse(query))
+func renderUtilization(metricData MetricData, extraLabelMatchers []string, query string) (string, error) {
+	// NOTE: Ideally we would use `github.com/prometheus/prometheus/promql/parser#ParseMetricSelector` here
+	// however the dependencies of the Prometheus server conflict.
 
-	// Always include the trial namespace
-	labels := []string{fmt.Sprintf("namespace=\"%s\"", data.Trial.Namespace)}
+	// Allow the extra label matchers to be variadic or comma-delimited
+	extraLabelMatchers = strings.Split(strings.Join(extraLabelMatchers, ","), ",")
 
-	for _, label := range strings.Split(strings.Join(labelArgs, ","), ",") {
-		if label == "" {
+	// Always include the trial namespace first
+	labelMatchers := make([]string, 0, 1+len(extraLabelMatchers))
+	labelMatchers = append(labelMatchers, fmt.Sprintf("namespace=\"%s\"", metricData.Trial.Namespace))
+	for _, labelMatcher := range extraLabelMatchers {
+		if labelMatcher == "" {
 			continue
 		}
 
-		kvpair := strings.Split(label, "=")
-		if len(kvpair) != 2 {
-			return "", fmt.Errorf("invalid label for utilization query, expected key=value, got: %s", label)
+		key, op, value, err := splitLabelMatcher(labelMatcher)
+		if err != nil {
+			return "", err
 		}
 
-		// Sanitize label name for prometheus usage
-		kvpair[0] = invalidLabelCharRE.ReplaceAllString(kvpair[0], "_")
-
-		labels = append(labels, fmt.Sprintf("label_%s=\"%s\"", kvpair[0], kvpair[1]))
+		labelMatchers = append(labelMatchers, key+op+value)
```

Comment:
If/when we take out the additional massaging of labels in `splitLabelMatcher`, will `splitLabelMatcher` turn more into a  validation step ( as in return bool ) and then we append `labelMatcher` directly to `labelMatchers` instead of reassembling it?

Reply:
I would prefer to make it a straight pass through and get rid of `splitLabelMatcher` altogether (just join the variadic args coming in from the template function using a comma; and even that is really just a concession to generalize making the metric selector optional). We can do some additional validation in the linter as you suggested to check the syntax.